File: Data Structures & Algorithms/time-based-key-value-store/submission-1.py
class TimeMap:

    def __init__(self):
        self.timemap = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.timemap:
            self.timemap[key] = []
        self.timemap[key].append([value,timestamp])

    def get(self, key: str, timestamp: int) -> str:
        values = self.timemap.get(key,[])

        left = 0
        right = len(values) - 1

        val = ''

        while left <= right:
            m = left + ((right - left) // 2)

            if timestamp < values[m][1]:
                right = m - 1
            elif timestamp >= values[m][1]:
                left = m + 1
                val = values[m][0]
            
            
        return val

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> myStack = new Stack<>();
        
        myStack.push(root);

        TreeNode curr = root;

        int c = 1;

        while (!myStack.isEmpty() || curr != null)
        {
            while(curr != null)
            {
                myStack.push(curr);
                curr = curr.left;
            }

            curr = myStack.pop();
            
            if (c == k) return curr.val;
            c++;

            curr = curr.right;
        }


        return root.val;
    }
}
